### PR TITLE
[Revisit] Use AzureDevOpsCredentialsProvider

### PR DESCRIPTION
**Pull request by awanlin**
## Hey, I just made a Pull Request!

Added support for using `AzureDevOpsCredentialsProvider` and deprecated `azureDevOps.token` configuration value.

This lays the foundation along with https://github.com/backstage/backstage/pull/19616 towards adding multi-org support to the Azure DevOps plugin

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
**Note:** https://github.com/backstage/backstage/pull/19620 closed on me and I pushed changes before trying to reopen it at which point it seems that it can't be reopened. This PR is the same starting code, I haven't made any changes based on feedback but will start doing that soon

**Comment by awanlin** [comment]
Hi @sanderaernouts, can you give this a look, please? As I mentioned I had to create a new PR for this. I've handled the following comments:

- Added a factory: https://github.com/backstage/backstage/pull/19620#discussion_r1307091600
- Updated to use the proper `authHandler`: https://github.com/backstage/backstage/pull/19620#discussion_r1308305491

But I think just another full review would be great. Once you are happy and approve we can get the core team to review 🚀

**Review by sanderaernouts — COMMENTED** [review]
@awanlin some small comments but otherwise LGTM. 

I did have a quick look at the `azure-devops-node-api` package, and I wonder if it would be cleaner to implement an `IRequestHandler` that deals with the authentication by invoking the `AzureCredentialsProvider`. That way, you wouldn't have to re-instantiate the `WebApi` every time.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (sanderaernouts) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:99]: I came accross [getHandlerFromToken](https://github.com/microsoft/azure-devops-node-api/blob/55908ae28d367e7b30c167f679c4f1087f647cb2/api/WebApi.ts#L67) that figures out whether its a bearer token or a PAT.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: Yes, good call, I've made this change and it's much nicer now 👍

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:103]: I played around with this but found it made the code messy. Until we removed the `azureDevOps.token` config value I think it's best to leave it as is. I did add a TODO about using this when that happens 👍

**Comment by sanderaernouts** [comment]
> > @awanlin some small comments but otherwise LGTM.
> > I did have a quick look at the `azure-devops-node-api` package, and I wonder if it would be cleaner to implement an `IRequestHandler` that deals with the authentication by invoking the `AzureCredentialsProvider`. That way, you wouldn't have to re-instantiate the `WebApi` every time.
> 
> I'm not sure I see how that helps? I need to build a new `WebApi` each time as there's no way for me to tell it to use a different authHandler and default server URL. Each call to the backend could be to get data from a different host + org combination. A user might be looking at an Entity for Cloud at Org A but then change entities and want the same info for On-Prem at Org Y.
> 
> I'm open to changes just feel boxed in by what I can do with this. Maybe we should chat on Discord about this?

I see your point, my idea was to have the `authHandler` figure that out based on the request URL. I didn't notice that the server URL was passed into the `WebApi` forcing you to create a new instance on each call.

**Review by sanderaernouts — APPROVED** [review]
@awanlin LGTM. One question though, aren't there any tests that require changing because of the `DefaultAzureDevOpsCredentialsProvider`?

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (vinzscam) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:108]: at this point `credentials` might be `undefined`. I'm not familiar with the logic behind credentials so I'm not sure whether we need some extra error handling here 🤔

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:73]:
```suggestion
  private readonly credentialsProvider: AzureDevOpsCredentialsProvider;
```

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:82]: these are good to do in fromConfig, since that leaves the constructor with just basic assignments, and also lets the constructor more easily receive higher-level mocks in testing

Comment (freben) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:82]: speaking of which, don't the actual api calls of this client ever get exercised by tests? 🤔

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:95]: maybe safest to encode the org here

File: plugins/azure-devops-backend/api-report.md
Comment (awanlin) [reply plugins/azure-devops-backend/api-report.md:27]: Yes, that makes sense to me. I've done that 👍

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: Yeah, I think when I revisited this from prior comments I didn't have a clear model of this in my head. The latest refactoring handles this much better IMO and ended up easier to read 👍

File: plugins/azure-devops-backend/src/service/router.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/service/router.ts:191]: Good call out, for now this will have to stay. The plan is to follow up with PR(s) to get this passed in instead, just setting the ground work in this PR

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: Good suggestion, done 👍

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: I've moved these to `fromConfig`. As for tests, I didn't have any when I initially contributed the plugin. I'll work on adding some 👍

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (sanderaernouts) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:123]: Alternatively, would a `codemod` be an option to move the `azureDevOps.token` token value to `integrations.azure.credentials[i].personalAccessToken`?

File: plugins/azure-devops-backend/src/service/router.ts
Comment (sanderaernouts) [reply plugins/azure-devops-backend/src/service/router.ts:191]: `azureDevOps.host` and `azureDevOps.organization` are deprecated now right? What happens when a user decides to remove the deprecated settings?

Comment (sanderaernouts) [reply plugins/azure-devops-backend/src/service/router.ts:191]: Maybe `azureDevOps.host` and `azureDevOps.organization` should not be deprecated just yet? Since you plan to change the route in the next PR(s) these settings can be deprecated as part of those?

**Review by sanderaernouts — CHANGES_REQUESTED** [review]

File: .changeset/unlucky-clouds-build.md
Comment (awanlin) [reply .changeset/unlucky-clouds-build.md]: Good catch, fixed 👍

File: plugins/azure-devops-backend/src/service/router.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/service/router.ts:191]: Alright, I've removed deprecating the config and removed the comment about that from the changeset.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:123]: My thinking on the final state of this once all the changes are in place is that if we can't find the information we need from the `integration.azure` config then we should throw an error with details as to how to correct it. There is simply no reasonable way we can provide a sound default across an unknown number of Orgs and/or Cloud vs Server setups. 

Now when it comes to the code in flight I'm trying to straddle these two worlds. Supporting what we have currently while also supporting the new `AzureDevOpsCredentialsProvider` so that we can migrate away from the `azureDevOps` config, which only ever supported PATs.

Ideally by the time this all lands Adopters would have already migrated to the new config introduced with the `AzureDevOpsCredentialsProvider` nearly 3 months ago. Given that I'm not sure there's going to be a lot of benefit of creating a `codemod` for this.

@sanderaernouts does that make sense?

Comment (sanderaernouts) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:123]: @awanlin makes sense 👍

**Review by sanderaernouts — APPROVED** [review]
@awanlin LGTM

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: Tests have been added 🚀

**Comment by awanlin** [comment]
@freben and @vinzscam this is good to go, would very much like to get this merged and into the next release. With this merged I can begin the work on the last step that should finish adding multi-org support 🚀

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:107]: Nobody calls this ever with a given host and org right? i presume there's some plan here

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:113]: `encodeURI` is almost never what you want, most commonly it's `encodeURIComponent`. But it shouldn't be harmful in this particular usage I think 🤔

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (freben) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:129]: Would it make sense to reconstruct this class and the handlers at most once every N seconds or something like that? Or is there no significant risk / performance impact of reconstructing it over and over for every request?

**Review by freben — APPROVED** [review]
Just optional / future questions, happy to approve

**Review by vinzscam — APPROVED** [review]

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:107]: Yes, the plan to tackle adding multi-org support was to do it in chunks - frontend, then backend which is this PR, and then a last PR to wire this all up between. This has made it easier to focus on the specific changes and made the PRs smaller

**Comment by awanlin** [comment]
Thanks @sanderaernouts @freben @vinzscam 🚀

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:113]: https://github.com/backstage/backstage/pull/21795 addresses this 👍

**Pull request merged**
